// ethereumjs-ledger/output/source/errors.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var ErrorCode;
(function (ErrorCode) {
    ErrorCode[ErrorCode["DeviceNotFound"] = 0] = "DeviceNotFound";
    ErrorCode[ErrorCode["EthereumAppNotOpen"] = 1] = "EthereumAppNotOpen";
    ErrorCode[ErrorCode["WrongMode"] = 2] = "WrongMode";
    ErrorCode[ErrorCode["InvalidInput"] = 3] = "InvalidInput";
    ErrorCode[ErrorCode["BadRequest"] = 4] = "BadRequest";
    ErrorCode[ErrorCode["Unknown"] = 5] = "Unknown";
})(ErrorCode = exports.ErrorCode || (exports.ErrorCode = {}));
var ErrorWithCode = (function (_super) {
    __extends(ErrorWithCode, _super);
    function ErrorWithCode(message, code) {
        var _this = _super.call(this, message) || this;
        Object.setPrototypeOf(_this, ErrorWithCode.prototype);
        _this.code = code;
        return _this;
    }
    return ErrorWithCode;
}(Error));
exports.ErrorWithCode = ErrorWithCode;
var WrappedError = (function (_super) {
    __extends(WrappedError, _super);
    function WrappedError(message, data) {
        var _this = _super.call(this, message, ErrorCode.Unknown) || this;
        _this.data = data;
        return _this;
    }
    return WrappedError;
}(ErrorWithCode));
exports.WrappedError = WrappedError;
